src/pudl/analysis/ferc1_eia_train.py

```diff
@@ -0,0 +1,712 @@
+
+    Args:
+        pudl_out (PudlTabl): the pudl_out object generated in a notebook and passed in.
+        util_dict (dict): A dictionary with keys that are the names of utility
+            parent companies and values that are lists of subsidiary utility_id_eia
+            values. EIA values are used instead of PUDL in this case because PUDL values
+            are subject to change.
+        years (list): A list of the years you'd like to add to the override sheets.
+        output_dir_path (str): The relative path to the folder where you'd like to
+            output the override spreadsheets that this function creates.
+    """
+    # Generate full input tables
+    inputs_dict = _generate_input_dfs(pudl_out)
+
+    # For each utility, make an override sheet with the correct input table slices
+    for util_name, util_id_eia_list in util_dict.items():
+        logger.info(f"Developing outputs for {util_name}")
+        util_year_subset_dict = _get_util_year_subsets(
+            inputs_dict, util_id_eia_list, years
+        )
+        _output_override_spreadsheet(util_year_subset_dict, util_name, output_dir_path)
+
+
+# --------------------------------------------------------------------------------------
+# Validate Overrides and Merge With Existing Training Data
+# --------------------------------------------------------------------------------------
+
+
+def _check_id_consistency(
+    id_col: Literal["record_id_eia_override_1", "record_id_ferc1"],
+    df,
+    actual_ids,
+    error_message,
+) -> None:
+    """Check for rogue FERC or EIA ids that don't exist.
+
+    Args:
+        id_col (str): The name of either the ferc record id column: record_id_ferc1 or
+            the eia record override column: record_id_eia_override_1.
+        df (pd.DataFrame): A dataframe of intended overrides.
+        actual_ids (list): A list of the ferc or eia ids that are valid and come from
+            either the ppl or official ferc-eia record linkage.
+        error_message (str): A short string to indicate the type of error you're
+            checking for. This could be looking for values that aren't in the official
+            list or values that are already in the training data.
+    """
+    logger.debug(f"Checking {id_col} consistency for {error_message}")
+
+    assert (
+        len(bad_ids := df[~df[id_col].isin(actual_ids)][id_col].to_list()) == 0
+    ), f"{id_col} {error_message}: {bad_ids}"
+
+
+def already_in_training(training_data, validated_connections):
+    """Blah."""
+    logger.info("checking if already in validation or not")
+    training_vals = training_data.record_id_eia.dropna().unique().tolist()
+    val_vals = validated_connections.record_id_eia_override_1.dropna().unique().tolist()
+
+    not_in_training = [x for x in val_vals if x not in training_vals]
+    print(f"Total records: {len(val_vals)}")
+    print(f"Records not in training data: {len(not_in_training)}")
+    print(not_in_training)
+    print("")
```

Comment:
what does this actually produce? in the notebook it just spits out `not_in_training` without a real indication of what those records are or what I'm supposed to do with them

Reply:
I'll clarify the logger -- it was a last minute thing I added so that I could tell if our manual mapping spreadsheets had already been integrated into the training data or not. Mostly an artifact of this process taking months / years to complete and old mapping files going stale.

Reply:
Actually, I clarified the doc string and removed it from the notebook, because it's only really necessary on specific occasions (when you forget what you've done...)